bindings/rust/bench/src/bin/memory.rs

```diff
@@ -0,0 +1,68 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use bench::{
+    harness::ConnectedBuffer, CipherSuite, CryptoConfig, ECGroup, OpenSslHarness, RustlsHarness,
+    S2NHarness, TlsBenchHarness,
+};
+use std::{fs::create_dir_all, path::Path};
+
+fn memory_bench<T: TlsBenchHarness>(dir_name: &str) {
+    println!("testing {dir_name}");
+
+    if !Path::new(&format!("target/memory/{dir_name}")).is_dir() {
+        create_dir_all(format!("target/memory/{dir_name}")).unwrap();
+    }
+
+    let mut harnesses = Vec::new();
+    harnesses.reserve(100);
+
+    // reserve space for buffers before benching
+    let mut buffers = Vec::new();
+    buffers.reserve(100);
+    for _ in 0..100 {
+        buffers.push(ConnectedBuffer::new());
+    }
+
+    // handshake one harness to initalize libraries
+    let mut harness = T::default().unwrap();
+    harness.handshake().unwrap();
+
+    // tell massif to take initial memory snapshot
+    crabgrind::monitor_command(format!("snapshot target/memory/{dir_name}/0.snapshot")).unwrap();
+
+    // make and handshake 100 harness
+    for i in 1..101 {
```

Comment:
I'm not sure I understand why 100 handshakes are needed. Is it only to compare one implementation to itself? If the same handshake is performed each time, why would `get_bytes_diff` change between each of them?

Reply:
In reality, `get_bytes_diff()` does change between them a little bit, especially for the first few handshakes or so. Even though one handshake is already performed before memory benching to init the libraries, the first few after that are also a little different than the rest (within ~5% I'm pretty sure). The handshakes after the first few sometimes vary by within 10 bytes but still vary. To get as accurate results as possible, the benches run 100 handshakes to smooth out the data.